File: myutils/src/main/java/com/jty/myutils/utils/DeviceUtils.java
package com.jty.myutils.utils;

import android.annotation.SuppressLint;
import android.content.Context;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;
import android.os.Build;
import android.provider.Settings;

import java.io.BufferedReader;
import java.io.Closeable;
import java.io.DataOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.NetworkInterface;
import java.util.Collections;
import java.util.List;

/**
 * Created by JTY on 2017/3/15 0015
 * 如需获取设备相关，请Utils.init();
 * 本篇感谢http://blankj.com 布兰柯鸡大神
 */

public class DeviceUtils {
    private DeviceUtils(){
        throw new UnsupportedOperationException("cannot be instantiated");
    }

    /**
     * 判断设备是否root
     */
    public static boolean isDeviceRooted() {
        String su = "su";
        String[] locations = {"/system/bin/", "/system/xbin/", "/sbin/", "/system/sd/xbin/", "/system/bin/failsafe/",
                "/data/local/xbin/", "/data/local/bin/", "/data/local/"};
        for (String location : locations) {
            if (new File(location + su).exists()) {
                return true;
            }
        }
        return false;
    }

    /**
     * 获取设备系统版本号
     * @return
     */
    public static int getSDKVersion() {
        return android.os.Build.VERSION.SDK_INT;
    }


    /**
     * 获取设备AndroidID
     *
     * @return AndroidID
     */
    @SuppressLint("HardwareIds")
    public static String getAndroidID() {
        return Settings.Secure.getString(Utils.getContext().getContentResolver(), Settings.Secure.ANDROID_ID);
    }

    /**
     * 获取设备MAC地址
     * <p>需添加权限 {@code <uses-permission android:name="android.permission.ACCESS_WIFI_STATE"/>}</p>
     * <p>需添加权限 {@code <uses-permission android:name="android.permission.INTERNET"/>}</p>
     *
     * @return MAC地址
     */
    public static String getMacAddress() {
        String macAddress = getMacAddressByWifiInfo();
        if (!"02:00:00:00:00:00".equals(macAddress)) {
            return macAddress;
        }
        macAddress = getMacAddressByNetworkInterface();
        if (!"02:00:00:00:00:00".equals(macAddress)) {
            return macAddress;
        }
        macAddress = getMacAddressByFile();
        if (!"02:00:00:00:00:00".equals(macAddress)) {
            return macAddress;
        }
        return "please open wifi";
    }

    /**
     * 获取设备MAC地址
     * <p>需添加权限 {@code <uses-permission android:name="android.permission.ACCESS_WIFI_STATE"/>}</p>
     *
     * @return MAC地址
     */
    @SuppressLint("HardwareIds")
    private static String getMacAddressByWifiInfo() {
        try {
            WifiManager wifi = (WifiManager) Utils.getContext().getSystemService(Context.WIFI_SERVICE);
            if (wifi != null) {
                WifiInfo info = wifi.getConnectionInfo();
                if (info != null) return info.getMacAddress();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return "02:00:00:00:00:00";
    }

    /**
     * 获取设备MAC地址
     * <p>需添加权限 {@code <uses-permission android:name="android.permission.INTERNET"/>}</p>
     *
     * @return MAC地址
     */
    private static String getMacAddressByNetworkInterface() {
        try {
            List<NetworkInterface> nis = Collections.list(NetworkInterface.getNetworkInterfaces());
            for (NetworkInterface ni : nis) {
                if (!ni.getName().equalsIgnoreCase("wlan0")) continue;
                byte[] macBytes = ni.getHardwareAddress();
                if (macBytes != null && macBytes.length > 0) {
                    StringBuilder res1 = new StringBuilder();
                    for (byte b : macBytes) {
                        res1.append(String.format("%02x:", b));
                    }
                    return res1.deleteCharAt(res1.length() - 1).toString();
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return "02:00:00:00:00:00";
    }

    /**
     * 获取设备MAC地址
     *
     * @return MAC地址
     */
    private static String getMacAddressByFile() {
        ShellUtils.CommandResult result = ShellUtils.execCmd("getprop wifi.interface", false);
        if (result.result == 0) {
            String name = result.successMsg;
            if (name != null) {
                result = ShellUtils.execCmd("cat /sys/class/net/" + name + "/address", false);
                if (result.result == 0) {
                    if (result.successMsg != null) {
                        return result.successMsg;
                    }
                }
            }
        }
        return "02:00:00:00:00:00";
    }

    /**
     * 获取设备厂商
     * @return
     */
    public static String getManufacturer() {
        return Build.MANUFACTURER;
    }

    /**
     * 获取设备型号
     * @return
     */
    public static String getModel() {
        String model = Build.MODEL;
        if (model != null) {
            model = model.trim().replaceAll("\\s*", "");
        } else {
            model = "";
        }
        return model;
    }


    /**
     * <pre>
     *     author: Blankj
     *     blog  : http://blankj.com
     *     time  : 2016/8/7
     *     desc  : Shell相关工具类
     *
     *     execCmd: 是否是在root下执行命令
     * </pre>
     */
    public static class ShellUtils {

        private ShellUtils() {
            throw new UnsupportedOperationException("u can't instantiate me...");
        }

        /**
         * 是否是在root下执行命令
         *
         * @param command 命令
         * @param isRoot  是否需要root权限执行
         * @return CommandResult
         */
        public static CommandResult execCmd(String command, boolean isRoot) {
            return execCmd(new String[]{command}, isRoot, true);
        }

        /**
         * 是否是在root下执行命令
         *
         * @param commands 多条命令链表
         * @param isRoot   是否需要root权限执行
         * @return CommandResult
         */
        public static CommandResult execCmd(List<String> commands, boolean isRoot) {
            return execCmd(commands == null ? null : commands.toArray(new String[]{}), isRoot, true);
        }

        /**
         * 是否是在root下执行命令
         *
         * @param commands 多条命令数组
         * @param isRoot   是否需要root权限执行
         * @return CommandResult
         */
        public static CommandResult execCmd(String[] commands, boolean isRoot) {
            return execCmd(commands, isRoot, true);
        }

        /**
         * 是否是在root下执行命令
         *
         * @param command         命令
         * @param isRoot          是否需要root权限执行
         * @param isNeedResultMsg 是否需要结果消息
         * @return CommandResult
         */
        public static CommandResult execCmd(String command, boolean isRoot, boolean isNeedResultMsg) {
            return execCmd(new String[]{command}, isRoot, isNeedResultMsg);
        }

        /**
         * 是否是在root下执行命令
         *
         * @param commands        命令链表
         * @param isRoot          是否需要root权限执行
         * @param isNeedResultMsg 是否需要结果消息
         * @return CommandResult
         */
        public static CommandResult execCmd(List<String> commands, boolean isRoot, boolean isNeedResultMsg) {
            return execCmd(commands == null ? null : commands.toArray(new String[]{}), isRoot, isNeedResultMsg);
        }

        /**
         * 是否是在root下执行命令
         *
         * @param commands        命令数组
         * @param isRoot          是否需要root权限执行
         * @param isNeedResultMsg 是否需要结果消息
         * @return CommandResult
         */
        public static CommandResult execCmd(String[] commands, boolean isRoot, boolean isNeedResultMsg) {
            int result = -1;
            if (commands == null || commands.length == 0) {
                return new CommandResult(result, null, null);
            }
            Process process = null;
            BufferedReader successResult = null;
            BufferedReader errorResult = null;
            StringBuilder successMsg = null;
            StringBuilder errorMsg = null;
            DataOutputStream os = null;
            try {
                process = Runtime.getRuntime().exec(isRoot ? "su" : "sh");
                os = new DataOutputStream(process.getOutputStream());
                for (String command : commands) {
                    if (command == null) continue;
                    os.write(command.getBytes());
                    os.writeBytes("\n");
                    os.flush();
                }
                os.writeBytes("exit\n");
                os.flush();
                result = process.waitFor();
                if (isNeedResultMsg) {
                    successMsg = new StringBuilder();
                    errorMsg = new StringBuilder();
                    successResult = new BufferedReader(new InputStreamReader(process.getInputStream(), "UTF-8"));
                    errorResult = new BufferedReader(new InputStreamReader(process.getErrorStream(), "UTF-8"));
                    String s;
                    while ((s = successResult.readLine()) != null) {
                        successMsg.append(s);
                    }
                    while ((s = errorResult.readLine()) != null) {
                        errorMsg.append(s);
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                closeIO(os, successResult, errorResult);
                if (process != null) {
                    process.destroy();
                }
            }
            return new CommandResult(
                    result,
                    successMsg == null ? null : successMsg.toString(),
                    errorMsg == null ? null : errorMsg.toString()
            );
        }

        /**
         * 关闭IO
         *
         * @param closeables closeables
         */
        public static void closeIO(Closeable... closeables) {
            if (closeables == null) return;
            for (Closeable closeable : closeables) {
                if (closeable != null) {
                    try {
                        closeable.close();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            }
        }

        /**
         * 返回的命令结果
         */
        public static class CommandResult {
            /**
             * 结果码
             **/
            public int    result;
            /**
             * 成功信息
             **/
            public String successMsg;
            /**
             * 错误信息
             **/
            public String errorMsg;

            public CommandResult(int result, String successMsg, String errorMsg) {
                this.result = result;
                this.successMsg = successMsg;
                this.errorMsg = errorMsg;
            }
        }
    }


}

File: mykotlinutils/src/main/java/com/jty/mykotlinutils/utils/DeviceUtils.kt
package com.jty.mykotlinutils.utils

import android.annotation.SuppressLint
import android.content.Context
import android.net.wifi.WifiInfo
import android.net.wifi.WifiManager
import android.os.Build
import android.provider.Settings

import java.io.File
import java.net.NetworkInterface
import java.util.Collections

/**
 * Created by TaoYuan on 2017/9/20 0020.
 *
 * Get Device Info
 */

class DeviceUtils private constructor() {
    init {
        throw UnsupportedOperationException("cannot be instantiated")
    }

    companion object {

        /**
         * Judge Device isRoot
         */
        val isDeviceRooted: Boolean
            get() {
                val su = "su"
                val locations = arrayOf("/system/bin/", "/system/xbin/", "/sbin/", "/system/sd/xbin/", "/system/bin/failsafe/", "/data/local/xbin/", "/data/local/bin/", "/data/local/")
                for (location in locations) {
                    if (File(location + su).exists()) {
                        return true
                    }
                }
                return false
            }

        /**
         * Get Device OS SDK Version
         * @return SDK Version
         */
        val sdkVersion: Int
            get() = android.os.Build.VERSION.SDK_INT


        /**
         * Get Device AndroidID
         *
         * @return AndroidID
         */
        @SuppressLint("HardwareIds")
        fun getAndroidID(context: Context): String {
            return Settings.Secure.getString(context.applicationContext.contentResolver, Settings.Secure.ANDROID_ID)
        }

        /**
         * Get Device MAC Address
         *
         * Need Add  Permission
         * `<uses-permission android:name="android.permission.ACCESS_WIFI_STATE"/>`
         * `<uses-permission android:name="android.permission.INTERNET"/>`
         *
         * @param context
         * @return MAC Address
         */
        fun getMacAddress(context: Context): String {
            var macAddress = getMacAddressByWifiInfo(context)
            if ("02:00:00:00:00:00" != macAddress) {
                return macAddress
            }
            macAddress = macAddressByNetworkInterface
            return if ("02:00:00:00:00:00" != macAddress) {
                macAddress
            } else "please open wifi"
        }

        /**
         * Get Device MAC Address
         * Need Add  Permission`<uses-permission android:name="android.permission.ACCESS_WIFI_STATE"/>`
         *
         * @return MAC Address
         */
        @SuppressLint("HardwareIds")
        private fun getMacAddressByWifiInfo(context: Context): String {
            try {
                val wifi = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
                if (wifi != null) {
                    val info = wifi.connectionInfo
                    if (info != null) return info.macAddress
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }

            return "02:00:00:00:00:00"
        }

        /**
         * Get Device MAC Address
         *
         * Need Add  Permission
         * `<uses-permission android:name="android.permission.INTERNET"/>`
         *
         * @return MAC Address
         */
        private val macAddressByNetworkInterface: String
            get() {
                try {
                    val nis = Collections.list(NetworkInterface.getNetworkInterfaces())
                    for (ni in nis) {
                        if (!ni.name.equals("wlan0", ignoreCase = true)) continue
                        val macBytes = ni.hardwareAddress
                        if (macBytes != null && macBytes.size > 0) {
                            val res1 = StringBuilder()
                            for (b in macBytes) {
                                res1.append(String.format("%02x:", b))
                            }
                            return res1.deleteCharAt(res1.length - 1).toString()
                        }
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }

                return "02:00:00:00:00:00"
            }


        /**
         * Get Device Firm
         * @return
         */
        val manufacturer: String
            get() = Build.MANUFACTURER

        /**
         * Get Device Model
         * @return
         */
        val model: String
            get() {
                var model: String? = Build.MODEL
                if (model != null) {
                    model = model.trim { it <= ' ' }.replace("\\s*".toRegex(), "")
                } else {
                    model = ""
                }
                return model
            }
    }
}
